packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteHelpers.kt

```diff
@@ -0,0 +1,119 @@
+package expo.modules.sqlite
+
+import java.io.File
+import java.io.IOException
+import java.util.ArrayList
+
+@Throws(IOException::class)
+internal fun ensureDirExists(dir: File): File {
+  if (!dir.isDirectory) {
+    if (dir.isFile) {
+      throw IOException("Path '$dir' points to a file, but must point to a directory.")
+    }
+    if (!dir.mkdirs()) {
+      var additionalErrorMessage = ""
+      if (dir.exists()) {
+        additionalErrorMessage = "Path already points to a non-normal file."
+      }
+      if (dir.parentFile == null) {
+        additionalErrorMessage = "Parent directory is null."
+      }
+      throw IOException("Couldn't create directory '$dir'. $additionalErrorMessage")
+    }
+  }
+  return dir
+}
+
+internal fun pluginResultsToPrimitiveData(results: List<SQLiteModule.SQLitePluginResult>): List<Any> {
+  return results.map { convertPluginResultToArray(it) }
+}
+
+private fun convertPluginResultToArray(result: SQLiteModule.SQLitePluginResult): List<Any?> {
+  val rowsContent = result.rows.map { row ->
+    row.map { value ->
+      when (value) {
+        null -> null
+        is String -> value
+        is Boolean -> value
+        else -> (value as Number).toDouble()
+      }
+    }
+  }
+
+  return arrayListOf(
+      result.error?.message,
+      result.insertId.toInt(),
+      result.rowsAffected,
+      result.columns,
+      rowsContent
+  )
+}
+
+private fun isPragma(str: String): Boolean {
+  return startsWithCaseInsensitive(str, "pragma")
+}
+
+private fun isPragmaReadOnly(str: String): Boolean {
+  if (!isPragma(str)) {
+    return false
+  }
+  return !str.contains('=')
+}
+
+internal fun isSelect(str: String): Boolean {
+  return startsWithCaseInsensitive(str, "select") || isPragmaReadOnly(str)
+}
+
+internal fun isInsert(str: String): Boolean {
+  return startsWithCaseInsensitive(str, "insert")
+}
+
+internal fun isUpdate(str: String): Boolean {
+  return startsWithCaseInsensitive(str, "update")
+}
+
+internal fun isDelete(str: String): Boolean {
+  return startsWithCaseInsensitive(str, "delete")
+}
+
+// identify an "insert"/"select" query more efficiently than with a Pattern
+private fun startsWithCaseInsensitive(str: String, substr: String): Boolean {
+  var i = -1
+  val len = str.length
+  while (++i < len) {
+    val ch = str[i]
+    if (!Character.isWhitespace(ch)) {
+      break
+    }
+  }
+  var j = -1
+  val substrLen = substr.length
+  while (++j < substrLen) {
+    if (j + i >= len) {
+      return false
+    }
+    val ch = str[j + i]
+    if (Character.toLowerCase(ch) != substr[j]) {
+      return false
+    }
+  }
+  return true
```

Comment:
IMHO this piece of code is far too complicated.
I suggest using [`String#lowercase`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/lowercase.html) to achieve the similar, but far more readable code:
```java
  val strLowercased = str.lowercase()
  val substrLowercased = substr.lowercase()
  return strLowercased.startsWith(substrLowercased)
```
Or even better: [`String#startsWith`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/starts-with.html) accepts flag `ignoreCase`, so let's just use it 😊 
```suggestion
  return str.startsWith(substr, true)
```

Reply:
It won't be the same. We strip white characters at the start of the string.  Moreover, the `startWith` function has `ingoreCase` parameter.

Reply:
Yeah, you're right, but I saw that you've already discussed this fact and found that there's `String#trim` method that would fit just right in this place 🥳

Reply:
SOrry I've created another comments thread about the very same thing here 😅 
Don't hold a grudge against me, please 🙏

Reply:
rewritten this function to trimStart + startsWith. It was 2-4 times faster earlier